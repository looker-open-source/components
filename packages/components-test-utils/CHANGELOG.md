# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.10.4](https://github.com/looker-open-source/components/compare/v0.10.3...v0.10.4) (2021-01-25)

**Note:** Version bump only for package @looker/components-test-utils





## [0.10.3](https://github.com/looker-open-source/components/compare/v0.10.2...v0.10.3) (2021-01-25)

**Note:** Version bump only for package @looker/components-test-utils





## [0.10.2](https://github.com/looker-open-source/components/compare/v0.10.1...v0.10.2) (2021-01-22)


### Bug Fixes

* On build properly set package.json "files" key ([#1806](https://github.com/looker-open-source/components/issues/1806)) ([c81c4bb](https://github.com/looker-open-source/components/commit/c81c4bb625c58ede49957aad90ac8d9d7b2c4b79))





## [0.10.1](https://github.com/looker-open-source/components/compare/v0.10.0...v0.10.1) (2021-01-22)


### Bug Fixes

* Properly build ems/cjs/ts for easy deep-imports ([#1804](https://github.com/looker-open-source/components/issues/1804)) ([2a523cd](https://github.com/looker-open-source/components/commit/2a523cd70b079944376c6dc87c18202e05a97b01))





# [0.10.0](https://github.com/looker-open-source/components/compare/v0.9.29...v0.10.0) (2021-01-19)


### Bug Fixes

* Properly configure build for tree-shaking support ([9202b72](https://github.com/looker-open-source/components/commit/9202b72bc5fb99eb3e1af7d9e56f3dc15b1df2a4))





## [0.9.15]

### Changed

- `@looker/components-test-utils` helpers will no longer produce `globalStyles`
