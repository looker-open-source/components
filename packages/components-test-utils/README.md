# Test Utilities

Shared functions used for mocking and testing Looker UI Components.
