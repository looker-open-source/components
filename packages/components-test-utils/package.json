{
  "name": "@looker/components-test-utils",
  "license": "MIT",
  "version": "0.9.12",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "files": [
    "lib"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/looker-open-source/components",
    "directory": "packages/components-test-utils"
  },
  "publishConfig": {
    "access": "public"
  },
  "devDependencies": {
    "@looker/components-providers": "^0.9.12",
    "@testing-library/react": "^10.4.8",
    "@types/enzyme": "^3.10.5",
    "@types/jest": "^26.0.10",
    "@types/react": "^16.9.49",
    "@types/react-test-renderer": "^16.9.3",
    "enzyme": "^3.11.0",
    "jest-styled-components": "^6.3.4",
    "react": "^16.13.1",
    "react-test-renderer": "^16.13.1",
    "styled-components": "^4.4.1"
  },
  "peerDependencies": {
    "react": "^16.11"
  },
  "gitHead": "62febc162e3e45124f403e2c84fc741f68fe6714"
}
