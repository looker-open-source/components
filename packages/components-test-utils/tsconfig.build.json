{
  "extends": "../../tsconfig.build.json",
  "compilerOptions": {
    "rootDir": "./src",
    "outDir": "./lib"
  },
  "include": ["../../types", "src/**/*"]
}
