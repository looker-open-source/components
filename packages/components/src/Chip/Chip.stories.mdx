import { Chip } from '@looker/components'
import { Meta } from '@storybook/addon-docs'

<Meta component={Chip} title="Docs/Components/Content/Chip" />

# Chip

The `Chip` component is a compact element that can be set as `disabled` and has `onDelete` availabel, it can be triggered by clicking or `onKeyDown` backspace.

```jsx
;() => {
  const alertTrigger = () => alert('hello')
  return (
    <Space gap="u1">
      <Chip>first chip</Chip>
      <Chip>
        this is a very long text that shows the word wrapping in a chip
      </Chip>
      <Chip disabled>disabled chip</Chip>
      <Chip onDelete={alertTrigger}>deletable</Chip>
    </Space>
  )
}
```

## iconLabel

The `Chip` component provides the prop `iconLabel` to customize the icon label. Hover over the the `x` icon to see the tooltip.

```jsx
;() => {
  const alertTrigger = () => alert('hello')
  return (
    <Space gap="u1">
      <Chip iconLabel="remove chip" onDelete={alertTrigger}>
        hover the x
      </Chip>
      <Chip onDelete={alertTrigger}>hover the x</Chip>
    </Space>
  )
}
```

## maxWidth

The width of the `Chip` will accomodate its contents, by default. Setting a `maxWidth` will cause any overflow text to be truncated with ellipses.

```jsx
;() => {
  return (
    <Space gap="u1">
      <Chip maxWidth={150}>short</Chip>
      <Chip maxWidth={150}>
        Very long text inside the chip will be truncated
      </Chip>
    </Space>
  )
}
```
