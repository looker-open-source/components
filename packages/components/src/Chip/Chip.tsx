/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import {
  color,
  reset,
  layout,
  LayoutProps,
  space,
  SpaceProps,
  typography,
  CompatibleHTMLProps,
  TypographyProps,
} from '@looker/design-tokens'
import React, { ReactNode, forwardRef, Ref, useState } from 'react'
import styled from 'styled-components'
import { darken, rgba } from 'polished'
import { IconButton } from '../Button'
import { Text, TextProps } from '../Text'
import { TruncateProps, truncate } from '../Text/truncate'

export interface ChipProps
  extends CompatibleHTMLProps<HTMLSpanElement>,
    LayoutProps,
    SpaceProps,
    TruncateProps,
    TypographyProps {
  children: ReactNode
  disabled?: boolean
  focusVisible?: boolean
  onDelete?: () => void
}

const ChipStyle = styled.span<ChipProps>`
  ${reset}

  ${color}
  ${layout}
  ${space}
  ${typography}
  ${IconButton}{
    background-color: transparent;
    flex-shrink: 0;
  }

  align-items: center;
  background-color: ${({ theme }) => theme.colors.keySubtle};
  border-radius: 4px;
  color: ${({ theme }) => theme.colors.keyInteractive};
  display: flex;
  justify-items: center;

  &:hover {
    background-color: ${(props) => props.theme.colors.keyAccent};
  }

  &:focus {
    background-color: ${({ theme }) =>
      darken(0.1, theme.colors.keyAccent).toUpperCase()};
    outline: none;
  }

  ${({ focusVisible, theme: { colors } }) =>
    focusVisible && `box-shadow: 0 0 0 0.15rem ${rgba(colors.keyFocus, 0.25)};`}

  ${({ disabled, theme: { colors } }) =>
    disabled &&
    `color: ${colors.neutral};
      background-color: ${colors.neutralAccent};

      &:hover {
        background-color: ${colors.neutralAccent};
      }
    `}
`

const ChipLabel = styled(Text)<TextProps & TruncateProps>`
  ${truncate}
`

const ChipJSX = forwardRef((props: ChipProps, ref: Ref<HTMLSpanElement>) => {
  const {
    children,
    disabled,
    fontSize,
    onBlur,
    onDelete,
    onKeyUp,
    onKeyDown,
    truncate,
    ...restProps
  } = props

  const [isFocusVisible, setFocusVisible] = useState(false)

  const handleOnKeyUp = (event: React.KeyboardEvent<HTMLSpanElement>) => {
    setFocusVisible(true)
    onKeyUp && onKeyUp(event)
  }

  const handleOnBlur = (event: React.FocusEvent<HTMLSpanElement>) => {
    setFocusVisible(false)
    onBlur && onBlur(event)
  }

  const handleOnKeyDown = (event: React.KeyboardEvent<HTMLSpanElement>) => {
    setFocusVisible(false)
    if (event.key === 'Backspace') {
      onDelete && onDelete()
    }
    onKeyDown && onKeyDown(event)
    setFocusVisible(false)
  }

  const handleDelete = () => {
    if (!disabled) {
      onDelete && onDelete()
    }
    setFocusVisible(false)
  }

  return (
    <ChipStyle
      disabled={disabled}
      focusVisible={isFocusVisible}
      onBlur={handleOnBlur}
      onKeyDown={handleOnKeyDown}
      onKeyUp={handleOnKeyUp}
      ref={ref}
      tabIndex={disabled ? undefined : 0}
      {...restProps}
    >
      <ChipLabel fontSize={fontSize} truncate={truncate}>
        {children}
      </ChipLabel>
      {onDelete && !disabled && (
        <IconButton
          disabled={disabled}
          icon="Close"
          label="Delete"
          ml="xsmall"
          onClick={handleDelete}
          size="xxsmall"
        />
      )}
    </ChipStyle>
  )
})

ChipJSX.displayName = 'ChipJSX'

export const Chip = styled(ChipJSX)``

Chip.defaultProps = {
  fontSize: 'xsmall',
  fontWeight: 'semiBold',
  height: 28,
  maxWidth: 320,
  mb: 'xxsmall',
  minWidth: 44,
  px: 'xsmall',
  py: 'xxsmall',
  truncate: true,
}
