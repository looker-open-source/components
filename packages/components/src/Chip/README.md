# Chip

The `Chip` component can represent multiple interactive elements in the same area, such as a multiple values on a single filter or a list of email addresses.

```tsx
export default function Basic(props: ChipProps) {
  return <Chip {...props}>Basic</Chip>;
}
```

The `onClick` handler is called when when the user clicks anywhere on the `Chip` except for the x icon.
This icon appears when `onDelete` is defined and calls that intead when clicked.
`onDelete` is also called via the delete key. Neither callback is called when `disabled` is true.

```tsx
export default function ClickAndDelete() {
  const handleClick = () => alert('Clicked!');
  const handleDelete = () => alert('Deleted!');

  return (
    <Space>
      <Chip onClick={handleClick}>Click Me</Chip>
      <Chip disabled onClick={handleClick}>
        Click Me (nothing happens)
      </Chip>
      <Chip onClick={handleClick} onDelete={handleDelete}>
        Delete Me
      </Chip>
      <Chip disabled onClick={handleClick} onDelete={handleDelete}>
        Delete Me (nothing happens)
      </Chip>
    </Space>
  );
}
```

## MaxWidth

The width of the `Chip` will accomodate its contents, by default. Setting a `maxWidth` will cause any overflow text to be truncated with ellipses. The full text will show in a tooltip on hover.

```tsx
export default function MaxWidth() {
  return (
    <Space gap="u1">
      <Chip maxWidth={150}>short</Chip>
      <Chip maxWidth={150} onDelete={() => alert('deleted')}>
        Very long text inside the chip will be truncated
      </Chip>
    </Space>
  );
}
```

## Prefix

Text in the `prefix` will render before the children, in normal font-weight, with a colon after it.

```tsx
export default function Prefix() {
  return <Chip prefix="role">admin</Chip>;
}
```

## Removable

The following is an example of a collection of removable chips.

```tsx
export default function Removable() {
  const [values, setValues] = useState(['Cheddar', 'Gouda', 'Swiss']);
  function getDeleteHandler(item: string) {
    return () => {
      const newValues = values.filter(value => value !== item);
      setValues(newValues);
    };
  }
  return (
    <Space>
      {values.map(item => (
        <Chip onDelete={getDeleteHandler(item)} role="option" key={item}>
          {item}
        </Chip>
      ))}
    </Space>
  );
}
```

# IconLabel

The `Chip` component provides the prop `iconLabel` to customize the icon label. Hover over the the `x` icon to see the tooltip.

```tsx
export default function IconLabel() {
  const alertTrigger = () => alert('You click on the X');
  return (
    <Space gap="u1">
      <Chip iconLabel="remove chip" onDelete={alertTrigger}>
        hover the x
      </Chip>
      <Chip onDelete={alertTrigger}>hover the x</Chip>
    </Space>
  );
}
```
