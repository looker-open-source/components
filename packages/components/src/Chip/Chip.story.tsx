/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import { Story } from '@storybook/react/types-6-0'
import React, { useState } from 'react'
import { ChipButton } from '../ChipButton'
import { Space } from '../Layout'
import { Chip, ChipProps } from './Chip'

export default {
  component: Chip,
  title: 'Chip',
}

const Template: Story<ChipProps & { text: string }> = ({ text, ...args }) => (
  <Chip {...args}>{text}</Chip>
)

export const Basic = Template.bind({})
Basic.args = {
  text: 'Basic',
}

export const Disabled = Template.bind({})
Disabled.args = {
  disabled: true,
  text: 'Disabled',
}

export const Deletable = Template.bind({})
Deletable.args = {
  onDelete: () => {
    //
  },
  text: 'Deletable',
}

export const Prefix = Template.bind({})
Prefix.args = {
  prefix: 'role',
  text: 'admin',
}

export const ClickAndDelete = () => {
  const handleClick = () => alert('Clicked!')
  const handleDelete = () => alert('Deleted!')
  return (
    <Space>
      <Chip onClick={handleClick}>Click Me</Chip>
      <Chip disabled onClick={handleClick}>
        Click Me (nothing happens)
      </Chip>
      <Chip onClick={handleClick} onDelete={handleDelete}>
        Delete Me
      </Chip>
      <Chip disabled onClick={handleClick} onDelete={handleDelete}>
        Delete Me (nothing happens)
      </Chip>
    </Space>
  )
}

ClickAndDelete.parameters = {
  storyshots: { disable: true },
}

export const ChipButtons = () => {
  const handleClick = () => alert('Clicked!')
  const handleDelete = () => alert('Deleted!')
  return (
    <Space>
      <ChipButton onClick={handleClick}>Click Me</ChipButton>
      <ChipButton disabled onClick={handleClick}>
        Click Me (nothing happens)
      </ChipButton>
      <ChipButton onClick={handleClick} onDelete={handleDelete}>
        Delete Me
      </ChipButton>
      <ChipButton disabled onClick={handleClick} onDelete={handleDelete}>
        Delete Me (nothing happens)
      </ChipButton>
    </Space>
  )
}

ChipButtons.parameters = {
  storyshots: { disable: true },
}

export const Removable = () => {
  const [values, setValues] = useState(['Cheddar', 'Gouda', 'Swiss'])
  function getDeleteHandler(item: string) {
    return () => {
      const newValues = values.filter((value) => value !== item)
      setValues(newValues)
    }
  }
  return (
    <Space>
      {values.map((item) => (
        <Chip onDelete={getDeleteHandler(item)} role="option" key={item}>
          {item}
        </Chip>
      ))}
    </Space>
  )
}

Removable.parameters = {
  storyshots: { disable: true },
}
