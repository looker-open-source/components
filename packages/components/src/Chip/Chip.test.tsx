/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */
import { assertSnapshot, renderWithTheme } from '@looker/components-test-utils'
import React from 'react'
import { fireEvent } from '@testing-library/react'
import { Chip } from './Chip'

test('Chip renders correctly', () => {
  assertSnapshot(<Chip>chip</Chip>)
})

test('Chip renders disabled', () => {
  assertSnapshot(<Chip disabled>chip</Chip>)
})

test('Chip accepts a prefix and renders it with correct style', () => {
  const { getByText } = renderWithTheme(<Chip prefix="role">admin</Chip>)
  expect(getByText(/\brole\b/)).toHaveStyleRule('font-weight: 400')
})

test('Chip onDelete renders correctly', () => {
  const onDeleteTrigger = jest.fn()

  const { getByTestId, getByRole } = renderWithTheme(
    <Chip onDelete={onDeleteTrigger} data-testid="chip">
      clickable
    </Chip>
  )

  fireEvent.click(getByRole('button'))
  expect(onDeleteTrigger).toHaveBeenCalledTimes(1)

  const chip = getByTestId('chip')

  fireEvent.keyDown(chip, {
    key: 'Backspace',
  })
  expect(onDeleteTrigger).toHaveBeenCalledTimes(2)
})
