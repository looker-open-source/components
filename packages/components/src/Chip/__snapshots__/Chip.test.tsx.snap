// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip renders correctly 1`] = `
.c0 {
  font-family: 'Roboto','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #F3F2FF;
  border-radius: 4px;
  color: #7a55e3;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: 0.75rem;
  font-weight: 600;
  height: 28px;
  max-width: 320px;
  min-width: 44px;
  padding: 0.25rem 0.5rem;
}

.c0:hover,
.c0:active,
.c0:focus,
.c0[aria-selected='true'] {
  background: #E8E5FF;
}

.c0.focus,
.c0:focus {
  outline: none;
}

.c0:active {
  border-color: #6C43E0;
}

.c0[disabled] {
  background: #F5F6F7;
  color: #71767a;
}

.c0[disabled]:hover {
  background: #F5F6F7;
}

.c1 {
  font-family: 'Roboto','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<span
  className="c0 "
  onBlur={[Function]}
  onClick={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  tabIndex={0}
>
  <span
    className="c1"
  >
    chip
  </span>
</span>
`;

exports[`Chip renders disabled 1`] = `
.c0 {
  font-family: 'Roboto','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #F3F2FF;
  border-radius: 4px;
  color: #7a55e3;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: 0.75rem;
  font-weight: 600;
  height: 28px;
  max-width: 320px;
  min-width: 44px;
  padding: 0.25rem 0.5rem;
}

.c0:hover,
.c0:active,
.c0:focus,
.c0[aria-selected='true'] {
  background: #E8E5FF;
}

.c0.focus,
.c0:focus {
  outline: none;
}

.c0:active {
  border-color: #6C43E0;
}

.c0[disabled] {
  background: #F5F6F7;
  color: #71767a;
}

.c0[disabled]:hover {
  background: #F5F6F7;
}

.c1 {
  font-family: 'Roboto','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<span
  className="c0 "
  disabled={true}
  onBlur={[Function]}
  onClick={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
>
  <span
    className="c1"
  >
    chip
  </span>
</span>
`;
